src/sockaddr.rs

```diff
@@ -12,35 +12,88 @@ use winapi::shared::ws2ipdef::SOCKADDR_IN6_LH_u;
 /// The address of a socket.
 ///
 /// `SockAddr`s may be constructed directly to and from the standard library
-/// `SocketAddr`, `SocketAddrV4`, and `SocketAddrV6` types.
+/// [`SocketAddr`], [`SocketAddrV4`], and [`SocketAddrV6`] types.
 pub struct SockAddr {
     storage: sockaddr_storage,
     len: socklen_t,
 }
 
+#[allow(clippy::len_without_is_empty)]
 impl SockAddr {
+    /// Create a `SockAddr` from the underlying storage and its length.
+    ///
+    /// The type of the address storage and length passed to the function `init`
+    /// is OS/architecture specific.
+    ///
+    /// # Safety
+    ///
+    /// - The `SockAddr`'s address family must match its contents.
+    /// - The given length of the `SockAddr` must be correct.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # fn main() -> std::io::Result<()> {
+    /// # #[cfg(unix)] {
+    /// use std::io;
+    /// use std::mem;
+    /// use std::os::unix::io::AsRawFd;
+    ///
+    /// use socket2::{SockAddr, Socket, Domain, Type};
+    ///
+    /// let socket = Socket::new(Domain::IPV4, Type::STREAM, None)?;
+    ///
+    /// // Initialise a `SocketAddr` byte calling `getsockname(2)`.
+    /// let mut addr_storage: libc::sockaddr_storage = unsafe { mem::zeroed() };
+    /// let mut len = 0;
```

Comment:
I think this should be `size_of_val(&addr_storage)` as otherwise the OS think it can only write zero bytes, and thus doesn't write the address.